File: src/bmsx/ecs/pipeline.ts
import type { World } from "../core/world";
import { ECSystem, TickGroup } from "./ecsystem";

declare const $: any; // avoid circular dependency issues

export interface SystemDescriptor {
	id: string;
	group: TickGroup;
	defaultPriority?: number;
	tags?: string[];
	create: (priority: number) => ECSystem;
}

export interface NodeSpec {
	/** Registered system id to include */
	ref: string;
	/** Override group */
	group?: TickGroup;
	/** Override priority */
	priority?: number;
	/** Enable by predicate */
	when?: (world: World) => boolean;
}

type NodeResolved = Required<Pick<NodeSpec, "ref">> & {
	group: TickGroup;
	priority: number;
	index: number; // insertion order for stability
};

export interface BuildDiagnostics {
	finalOrder: string[];
	groupOrders: Record<number, string[]>; // key: TickGroup
	buildMs: number;
}

export class ECSPipelineRegistry {
	private _descs = new Map<string, SystemDescriptor>();
	private _lastDiagnostics: BuildDiagnostics | null = null;

	register(desc: SystemDescriptor): void {
		if (this._descs.has(desc.id)) throw new Error(`ECSPipelineRegistry: duplicate id '${desc.id}'`);
		this._descs.set(desc.id, desc);
	}

	registerMany(descs: SystemDescriptor[]): void { for (const d of descs) this.register(d); }

	get(id: string): SystemDescriptor | undefined { return this._descs.get(id); }

	/** Build and assign systems to the world's ECSystemManager based on nodes. */
	build(world: World, nodes: NodeSpec[]): BuildDiagnostics {
		const t0 = $.platform.clock.now();
		const filtered = nodes.filter(n => (n.when ? n.when(world) : true));
		const resolved: NodeResolved[] = [];
		for (let i = 0; i < filtered.length; i++) {
			const n = filtered[i];
			const d = this._descs.get(n.ref);
			if (!d) throw new Error(`ECSPipelineRegistry: unknown system ref '${n.ref}'`);
			resolved.push({
				ref: n.ref,
				group: n.group ?? d.group,
				priority: n.priority ?? d.defaultPriority ?? 0,
				index: i,
			});
		}

		// Sort by group, then priority, then index
		resolved.sort((a, b) => {
			if (a.group !== b.group) return a.group - b.group;
			if (a.priority !== b.priority) return a.priority - b.priority;
			return a.index - b.index;
		});

		const finalOrder: NodeResolved[] = resolved;
		const groupOrders: Record<number, string[]> = {};

		// Populate groupOrders for diagnostics
		for (const r of finalOrder) {
			if (!groupOrders[r.group]) groupOrders[r.group] = [];
			groupOrders[r.group].push(r.ref);
		}

		// Instantiate systems
		const systems: ECSystem[] = [];
		for (const r of finalOrder) {
			const d = this._descs.get(r.ref)!;
			const sys = d.create(r.priority);
			sys.__ecsId = r.ref;
			systems.push(sys);
		}

		// Apply to world
		world.systems.clear();
		for (const s of systems) world.systems.register(s);

		const t1 = $.platform.clock.now();
		const diag: BuildDiagnostics = {
			finalOrder: finalOrder.map(r => r.ref),
			groupOrders,
			buildMs: t1 - t0,
		};
		this._lastDiagnostics = diag;
		return diag;
	}

	getLastDiagnostics(): BuildDiagnostics | null { return this._lastDiagnostics; }
}

export const DefaultECSPipelineRegistry = new ECSPipelineRegistry();

File: src/bmsx/res/bios/ecs/pipeline.lua
-- ecs_pipeline.lua
-- ecs pipeline registry and builder for lua engine

local registry<const> = require('registry')

local ecspipelineregistry<const> = {}
ecspipelineregistry.__index = ecspipelineregistry

function ecspipelineregistry.new()
	local self<const> = setmetatable({}, ecspipelineregistry)
	self._descs = {}
	return self
end

function ecspipelineregistry:register(desc)
	self._descs[desc.id] = desc -- Allow overriding existing descs with the same id, to allow for dynamic changes to the pipeline and hot-resume.
end

function ecspipelineregistry:register_many(descs)
	for i = 1, #descs do
		self:register(descs[i])
	end
end

function ecspipelineregistry:get(id)
	return self._descs[id]
end

function ecspipelineregistry:build(world_instance, nodes)
	local filtered<const> = {}
	for i = 1, #nodes do
		local n<const> = nodes[i]
		if not n.when or n.when(world_instance) then
			filtered[#filtered + 1] = n
		end
	end

	local resolved<const> = {}
	for i = 1, #filtered do
		local n<const> = filtered[i]
		local d<const> = self._descs[n.ref]
		if not d then
			error('ecspipelineregistry: unknown system ref "' .. n.ref .. '"')
		end
		local create_priority = n.priority
		if create_priority == nil then
			create_priority = d.default_priority
		end
		resolved[#resolved + 1] = {
			ref = n.ref,
			group = n.group or d.group,
			priority = create_priority,
			create_priority = create_priority,
			index = i,
		}
	end

	table.sort(resolved, function(a, b)
		if a.group ~= b.group then
			return a.group < b.group
		end
		if a.priority ~= b.priority then
			return a.priority < b.priority
		end
		return a.index < b.index
	end)

	local systems<const> = {}
	for i = 1, #resolved do
		local r<const> = resolved[i]
		local d<const> = self._descs[r.ref]
		local sys<const> = d.create(r.create_priority)
		sys.__ecs_id = r.ref
		sys.id = 'ecs:' .. r.ref
		sys.type_name = 'ecsystem'
		systems[#systems + 1] = sys
	end

	-- Deregister previous ECS system instances from the registry.
	for i = 1, #world_instance.systems.systems do
		local old<const> = world_instance.systems.systems[i]
		if old.id then
			registry.instance:deregister(old.id, true)
		end
	end

	world_instance.systems:clear()
	for i = 1, #systems do
		world_instance.systems:register(systems[i])
		registry.instance:register(systems[i])
	end
	world_instance:rebind_subsystem_systems_all()
end

return {
	ecspipelineregistry = ecspipelineregistry,
	defaultecspipelineregistry = ecspipelineregistry.new(),
}
